### ISSUE TITLE
`file_server` no longer responds with 404 for missing files

### ISSUE THREAD

**Issue by Saklad5** [caddyserver/caddy#5059]
Caddy 2.6, in contrast to Caddy 2.5, seems to serve missing files with a 200 status code instead of a 404 code.

This can be observed on Caddy's own website: <https://caddyserver.com/docs> and <https://caddyserver.com/doc> both respond with 200, even though the latter doesn't exist and has no content.

This is an extremely serious bug, in my estimation, with the potential to seriously impair websites through inadvertent caching and skipped error handling. It should be fixed as soon as possible.

**Comment by mholt** [comment]
Hmm, no I definitely still get 404 with `caddy file-server` and then accessing pages that don't exist.

Something might be different, but it appears to be a niche case.

**Comment by Saklad5** [comment]
Hmm. Would you like me to upload my Caddyfile? I don't think I have anything that could impact it, but if you can't reproduce it, it may help.

**Comment by mholt** [comment]
@Saklad5 That would be great. Here's a template that might be overkill, but the more info the better:

## Template

<pre><code>## 1. Environment

### 1a. Operating system and version

```
paste here
```


### 1b. Caddy version (run `caddy version` or paste commit SHA)

```
paste here
```


### 1c. Go version (if building Caddy from source; run `go version`)

```
paste here
```


## 2. Description

### 2a. What happens (briefly explain what is wrong)




### 2b. Why it's a bug (if it's not obvious)




### 2c. Log output

```
paste terminal output or logs here
```



### 2d. Workaround(s)




### 2e. Relevant links




## 3. Tutorial (minimal steps to reproduce the bug)




</code></pre>

## Instructions -- please heed otherwise we cannot help you (help us help you!)

1. **Environment**: Please fill out your OS and Caddy versions, even if you don't think they are relevant. (They are _always_ relevant.) If you built Caddy from source, provide the commit SHA and specify your exact Go version.

2. **Description:** Describe at a high level what the bug is. What happens? Why is it a bug? Not all bugs are obvious, so convince readers that it's actually a bug.
	- **2c) Log output:** Paste terminal output and/or complete logs in a code block. [DO NOT REDACT INFORMATION except for credentials.](https://caddy.community/t/how-to-get-help-with-caddy-more-effectively/5222)
	- **2d) Workaround:** What are you doing to work around the problem in the meantime? This can help others who encounter the same problem, until we implement a fix.
	- **2e) Relevant links:** Please link to any related issues, pull requests, docs, and/or discussion. This can add crucial context to your report.

3. **Tutorial:** What are the _minimum required specific steps_ someone needs to take in order to experience the same bug? Your goal here is to make sure that anyone else can have the same experience with the bug as you do. You are writing a tutorial, so make sure to carry it out yourself before posting it. Please:
	- Start with an empty config. Add _only_ the lines/parameters that are _absolutely required_ to reproduce the bug.
	- Do not run Caddy inside containers.
	- Run Caddy manually in your terminal; do not use systemd or other init systems.
	- If making HTTP requests, avoid web browsers. Use a simpler HTTP client instead, like `curl`.
	- Do not redact any information from your config (except credentials). Domain names are public knowledge and [often necessary for quick resolution of an issue](https://caddy.community/t/how-to-get-help-with-caddy-more-effectively/5222)!
	- Note that ignoring this advice may result in delays, or even in your issue being closed. 😞 Only actionable issues are kept open, and if there is not enough information or clarity to reproduce the bug, then the report is not actionable.

Example of a tutorial:

<blockquote>Create a config file:

```
{ ... }
```

Open terminal and run Caddy:

```
$ caddy ...
```

Make an HTTP request:

```
$ curl ...
```

Notice that the result is ___ but it should be ___.</blockquote>

**Comment by FarisZR** [comment]
I can replicate on my blog

with 2.5, it shows the 404 page with 404 response code
on 2.6, it shows the page, but with 200 response code.

https://fariszr.com/en/tst/

**Comment by mholt** [comment]
@FarisZR Please post your config and any other information we'd need to reproduce this. Thanks

**Comment by Saklad5** [comment]
## 1. Environment

### 1a. Operating system and version

```
Ubuntu 18.04.6 LTS
```


### 1b. Caddy version (run `caddy version` or paste commit SHA)

```
v2.6.0 h1:lHDynvM+sTOi9Aq4Y15b4FtkqzPB36WbUrZvVdwzTCA=
```

---

```
{
        acme_ca https://acme-v02.api.letsencrypt.org/directory
        email admin@saklad5.com

        admin unix//run/caddy/caddy-admin.sock

        servers {
                protocols h1 h2 h2c
                strict_sni_host
        }
}

(global) {
        @https {
                protocol https
        }
        @unsupported {
                method POST PUT DELETE CONNECT OPTIONS TRACE
        }

        header @https ?Strict-Transport-Security "max-age=63072000; includeSubDomains"
        header ?Content-Security-Policy "base-uri 'none' default-src 'none'"
        header ?Cache-Control "max-age=600; stale-while-revalidate=85800; no-transform"
        encode zstd br gzip
        error @unsupported 501
}

(files) {
        file_server {
                root /var/www/{args.0}
        }
}

(onion) {
        header @https Alt-Svc "h2=\"{args.0}{env.ONION_ROOT}:443\"; persist=1"
        header @https Onion-Location "http://{args.0}{env.ONION_ROOT}{http.request.uri.path}"
}

saklad5.com http://{env.ONION_ROOT} {
        import global
        import files @
        import onion ""

        header @https Strict-Transport-Security "max-age=63072000; includeSubDomains; preload"
}

mta-sts.saklad5.com http://mta-sts.{env.ONION_ROOT} {
        import global
        import files {labels.2}
        import onion {labels.2}.
}
```
---
The nature of the issue and the steps to reproduce it were detailed earlier.

**Comment by mholt** [comment]
@Saklad5 Thanks. While I work on this, what is in your logs? Please enable `debug` mode in your global options and also the access log (`log` directive in your site block). That could help speed things up.

**Comment by FarisZR** [comment]
## 1. Environment

### 1a. Operating system and version

```
Official Caddy server alpine image, latest tag
```

### 1b. Caddy version (run `caddy version` or paste commit SHA)

```
v2.6.0 h1:lHDynvM+sTOi9Aq4Y15b4FtkqzPB36WbUrZvVdwzTCA=
```

### 2c. Log output

```
caddy-caddy-1  | {"level":"debug","ts":1663716162.574832,"logger":"http.handlers.rewrite","msg":"rewrote request","request":{"remote_ip":"xxxxxx","remote_port":"57966","proto":"HTTP/3.0","method":"GET","host":"fariszr.com","uri":"/favicon.ico","headers":{"Accept":["image/avif,image/webp,image/apng,image/svg+xml,image/*,*/*;q=0.8"],"Sec-Gpc":["1"],"Accept-Language":["de-DE,de;q=0.8"],"Sec-Fetch-Mode":["no-cors"],"Cache-Control":["no-cache"],"User-Agent":["Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36"],"Sec-Fetch-Dest":["image"],"Referer":["https://fariszr.com/en/tst/"],"Accept-Encoding":["gzip, deflate, br"],"Pragma":["no-cache"],"Sec-Fetch-Site":["same-origin"]},"tls":{"resumed":false,"version":0,"cipher_suite":0,"proto":"","server_name":""}},"method":"GET","uri":"/en/404.html"}
```



### 2e. Relevant links
https://fariszr.com/en/tst

### notes
I think i have discovery another bug, the 404 page shows HTML code when loaded over http3, but works fine on http2...

**Comment by mholt** [comment]
@FarisZR Can you please fill out the whole template? I don't really know what to do with that information.

It looks like Caddy is emitting a 404 according to the logger. I'm curious where it turns into a 200, but I need more info.

**Comment by FarisZR** [comment]
Updated

**Comment by FarisZR** [comment]
Here is repo where i could replicate this bug

Just clone it, and run it with compose, `docker compose up -d`.

https://gitlab.com/fzr-forks/caddy-test

I couldn't r replicate the HTTP/3 bug where the page would be HTML code instead of the actual page, because for some reason i couldn't get HTTP3 to work locally.

But you can replicate it on my website, the 404 page doesn't show up correctly when served over HTTP/3

**Comment by WeidiDeng** [comment]
@Saklad5 Can you download from [artifacts](https://github.com/caddyserver/caddy/actions/runs/3094361009#artifacts) to see if the problem is fixed?

**Comment by Saklad5** [comment]
> @Saklad5 @FarisZR Can you download from [artifacts](https://github.com/caddyserver/caddy/actions/runs/3094361009#artifacts) to see if the problem is fixed?

I'm quite new to Caddy, having only started using it a few weeks ago. Is there a one-liner I could use to test this on a different system? I don't want to use the artifact directly in production.

**Comment by francislavoie** [comment]
I'm not sure what you mean by a "one-liner". Download the artifact (Caddy binary build during CI) for your platform (see the bottom of that link, if you're on linux, download the `caddy_Linux_go1.19_cca7e5d` one), unzip it, backup your existing Caddy binary, put this new one in place of it, then run Caddy again and try the same things you were seeing a problem with before.

**Comment by Saklad5** [comment]
I can confirm that the artifact doesn't have the problem.

**Comment by FarisZR** [comment]
Weird, that fixed it for me
https://gitlab.com/fzr-forks/caddy-test/-/tree/fixed

Still trying to replicate the broken 404-page on http3, bug.
Looks like I have to use a valid domain/certificate for that.

**Comment by mholt** [comment]
Thank you both for confirming!

**Comment by FarisZR** [comment]
Is there an ARM64 binary from the patched version?

**Comment by mholt** [comment]
Thanks @Saklad5 and @FarisZR for confirming!

I believe @WeidiDeng also verified that this fixes the broken 404 page on http3 bug. (Based on comments in Slack)

**Comment by mholt** [comment]
Released in v2.6.1: https://github.com/caddyserver/caddy/releases/tag/v2.6.1
